libvast/src/system/index.cpp

```diff
@@ -771,6 +788,107 @@ index_state::status(status_verbosity v) const {
+            && active_partition.id == partition_id) {
+          part = active_partition.actor;
+          break;
+        }
+      }
+      if (!part) {
+        if (auto it = st.unpersisted.find(partition_id);
+            it != st.unpersisted.end())
+          part = it->second;
+        else if (auto it = st.persisted_partitions.find(partition_id);
+                 it != st.persisted_partitions.end())
+          part = st.inmem_partitions.get_or_load(partition_id);
+      }
+      if (!part)
+        VAST_ERROR("{} could not load partition {} that was part of a "
+                   "query",
+                   *st.self, partition_id);
+      return part;
+    };
+    auto partition_actor = materialize(next->partition);
+    if (!partition_actor) {
+      // for (auto qid : next.queries) {
+      //   auto& query_state = st.pending_queries[qid];
+      //   query_state.candidate_partitions--;
+      // }
+      continue;
+    }
+    st.counters.partition_materializations++;
+    st.counters.partition_lookups += next->queries.size();
+    // 3. request all relevant queries in a loop
+    auto cnt = std::make_shared<size_t>(next->queries.size());
+    for (auto qid : next->queries) {
+      auto& query_state = st.pending_queries.queries[qid];
+      auto handle_completion = [cnt, qid, &st] {
+        // TODO: use find
+        auto& query_state = st.pending_queries.queries[qid];
+        query_state.completed_partitions++;
+        if (query_state.completed_partitions
+            == query_state.requested_partitions) {
+          st.self->send(query_state.client, atom::done_v);
+          auto response = query_state.completed_partitions
+                              == query_state.candidate_partitions
+                            ? query_response::complete
+                            : query_response::incomplete;
+          query_state.rp.deliver(response);
+        }
+        // 4. Remove queries with no more remaining partitions from
+        // pending_queries.
+        if (query_state.completed_partitions
+            == query_state.candidate_partitions)
+          st.pending_queries.queries.erase(qid);
+        // 5. recursively call schedule_lookups in the done handler. ...or
+        //    when all done? (6)
+        // 6. decrement st.running_partition_lookups when all queries that
+        //    were started are done. Keep track in the closure.
+        *cnt -= 1;
+        if (*cnt == 0) {
+          --st.running_partition_lookups;
+          schedule_lookups(st);
+        }
+      };
+      st.self->request(partition_actor, caf::infinite, query_state.query)
+        .then(
+          [handle_completion, pid = next->partition, &st](uint64_t n) {
+            VAST_TRACE("{} received {} results from {}", *st.self, n, pid);
+            handle_completion();
+          },
+          [handle_completion, &st](const caf::error& err) {
+            VAST_WARN("{} failed in partition lookup: {}", *st.self, err);
```

Comment:
Can we write this in a way that is meaningful to the user, e.g., somehow tell the user that for the query X we failed to get results from partition Y because of error Z?